Distributor guide (#452)
lidofinance/docs: docs/guides/reward-distributor-bot.md

# Reward distribution bot

## Introduction

> Permissionless reward distribution bot for Lido staking modules. Operates with smart contract based on [Node Operator Registry](https://docs.lido.fi/contracts/node-operators-registry) smart contract. 

Reviewer: ```suggestion
Permissionless reward distribution bot for Lido staking modules. Operates with smart contract based on the [Node Operator Registry](/contracts/node-operators-registry) smart contract. 
```

> After [Accounting Oracle](https://docs.lido.fi/guides/oracle-spec/accounting-oracle) completes the third phase, anyone can initiate the reward distribution to allocate rewards among Node Operators in the Staking Module.

Reviewer: ```suggestion
After [Accounting Oracle](https://docs.lido.fi/guides/oracle-spec/accounting-oracle) completes the third phase, anyone can initiate the reward distribution to allocate rewards among Node Operators in the Staking Module unless the next oracle frame starts.
```

Author: ```suggestion
After the [Accounting Oracle](https://docs.lido.fi/guides/oracle-spec/accounting-oracle) completes the third phase, anyone can initiate reward distribution to allocate rewards among Node Operators in the Staking Module, unless the oracle sends the next frame report.
```

> After the [Accounting Oracle](https://docs.lido.fi/guides/oracle-spec/accounting-oracle) completes the third phase, anyone can initiate reward distribution to allocate rewards among Node Operators in the Staking Module, unless the oracle sends the next frame report.

Reviewer: ```suggestion
After the [Accounting Oracle](/guides/oracle-spec/accounting-oracle) completes the third phase, anyone can initiate reward distribution to allocate rewards among Node Operators in the Staking Module, unless the oracle sends the next frame report.
```

## Requirements

### Hardware

- 1-core CPU
- 2GB RAM

> ### Nodes

Reviewer: Does it need a node or just an RPC endpoint? Light node, full node, archive node?

Author: Whatever, no special requirements

> - Execution Node

Reviewer: ```suggestion
- Execution Node (can be an RPC)
```

## How to use

Every epoch daemon checks the staking modules provided in environment. If a module has a non-distributed rewards, the bot pulls the trigger and distributes rewards between Node Operators inside this module.

> Basically bot is watching [RewardDistributionState](https://docs.lido.fi/contracts/node-operators-registry#getrewarddistributionstate) of module. 

Reviewer: ```suggestion
Basically bot is watching [RewardDistributionState](/contracts/node-operators-registry#getrewarddistributionstate) of module. 
```

why it is important: internal links are verified upon building stage and broken links are visible in CI

> If module has `ReadyForDistribution` state, bot triggers [distributeReward](https://docs.lido.fi/contracts/node-operators-registry#distributereward) method to distribute rewards.

Reviewer: ```suggestion
If module has `ReadyForDistribution` state, bot triggers [distributeReward](/contracts/node-operators-registry#distributereward) method to distribute rewards.
```

### Envs

Required variables are:

> - WEB3_RPC_ENDPOINTS - list of EL rpc endpoints separated by `,`.

Reviewer: why need a list?

Author: ```suggestion
- WEB3_RPC_ENDPOINTS - list of EL rpc endpoints separated by `,`. All except the first rpc will serve as fallback options.
```

> - NODE_OPERATOR_REGISTRY_ADDRESSES - List of staking modules that can accept rewards distribution. All addresses could be found [here](https://docs.lido.fi/deployed-contracts/).

Reviewer: ```suggestion
- NODE_OPERATOR_REGISTRY_ADDRESSES - List of staking modules that can accept rewards distribution. All addresses could be found [here](/deployed-contracts/).
```
